### PR TITLE
Add torch.ops.out_dtype

### PR DESCRIPTION
**Pull request by angelayi**
https://docs.google.com/document/d/10DYFG2sU3TSvguFP5kYwYLlo45KHFg3BhBOkUk0NKsU/edit#bookmark=id.hgfzmhlzkamk

Renamed mixed_dtype --> out_dtype because "mixed_dtype is not very descriptive in the context of regular pytorch where we support type promotion on most ops"

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by jerryzh168** [comment]
I feel `output_dtype` name is a bit weird, especially when it also has an argument called out_dtype...is there any better alternatives?

**Comment by ezyang** [comment]
How about `compute_precision`?

File: functorch/experimental/output_dtype.py
Comment (zou3519) [inline functorch/experimental/output_dtype.py:25]: Let's put the implementation into `torch/_higher_order_ops/output_dtype.py` so we don't need to do the work of moving it out of functorch.experimental later. Dynamo has some interesting behavior on files in the torch namespace and not in the torch namespace.

File: functorch/experimental/output_dtype.py
Comment (zou3519) [inline functorch/experimental/output_dtype.py:107]: - For safety, we should probably assert in all implementations that op is a torch._ops.OpOverload. For regular Dispatcher operators, we do a parsing step when invoking the operator, but there's not really a good way to do that with HigherOrderOperator right now
- As written the functionalization rule is incorrect. Let's say `op` is `torch.sin_`: then the functionalization rule should invoke `output_dtype(torch.sin, out_dtype, *args)` at some point. Do you see situations where we will need `output_dtype` for in-place or view operations? For now, I would recommend just asserting that `op` has a functional schema, and not worry about the functionalization rule.

**Review by zou3519 — COMMENTED** [review]
This looks like a good start, I added some comments around testing and the implementation

**Comment by jerryzh168** [comment]
I checked with chatGPT as well, and considering with all the suggestions, maybe `promote_output_dtype_and_apply` or `cast_input_to_output_dtype_and_apply`? these are a bit long though

here is the original Q & A:
```
I'm defining a new higher order operator in PyTorch that takes an operator and an output dtype and the arguments of the operator, it casts all the input Tensors to use the output dtype and then call the operator, what can be the names for this higher order operator?
```
```
There are several possible names for this higher-order operator in PyTorch. Here are a few suggestions:

dtype_cast_and_apply: This name clearly conveys the functionality of the operator, indicating that it casts the input tensors to the specified output dtype and then applies the given operator.

type_converted_operator: This name emphasizes the conversion of tensor types before applying the operator.

dtype_transform_operator: This name highlights the transformation of the dtype for the input tensors before applying the operator.

type_adaptive_operator: This name suggests that the operator adapts to the specified output dtype by converting the input tensors accordingly.

type_preserving_apply: This name conveys that the operator preserves the specified output dtype by casting the input tensors before applying the operator.

Ultimately, the choice of name depends on your preference and the context in which the operator will be used. It's important to select a name that accurately reflects the functionality and purpose of the operator to make it clear to other developers who may use or interact with your code.
```

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [reply functorch/experimental/output_dtype.py:65]: Yeah so this is exactly what I would like to talk about. Take following example:
```
class M(torch.nn.Module):
  def __init__(self):
    self.linear1 = ...
    self.linear2 = ...

  def forward(self, x):
    return (self.linear2(self.linear1(x))
```

If I want to quantize M via quantization aware training, the current workflow on top of exported graph, will modify forward graph to effectively do
```
call this FakeQuantizeRepresentation
  def forward(self, x):
    x = fake_quantize(x, <some-quant-params-that-are-being-learnt-for-linear1>) # this step, effectively, just uses quant params to quantize and dequntize x
    x = self.linear1(x)
    x = fake_quantize(x, <some-quant-params-that-are-being-learnt-for-linear2>)
    x = self.linear2(x)
    return x
```

Above may not be entirely accurate, but close enough (@jerryzh168 @andrewor14 can confirm)

Fake quantize do not accurately capture numerical behavior of target hardware. While achieving bit-exact behavior might be hard, it will be very useful to do better approximation of numerics than what is done by fake quantize nodes. If we were able to do better approximation, then, the hypothesis is that, we will have computed loss that is more accurate, which may result in gradient updates that account for achievable accuracy on target hw. So what should `forward` look like with better numerics? We approximate that with integer compute e.g.

```
call this IntRepresentation
  def forward(self, x):
    x = quantize_int8(x, <some-quant-params-that-are-being-learnt-for-linear1>)
    x = self.int8_linear1(x, <some-quant-params-that-are-being-learnt-for-linear1>) # decomposition of this op uses "mixed_dtype" op introduces in this diff. Note that int8_linear will return int8 tensor
    x = self.int8_linear2(x, <some-quant-params-that-are-being-learnt-for-linear2>)
    return dequantize(x, <some-quant-params-that-are-being-learnt-for-linear2>)
```

Note that in IntRepresentation linear1 and linear2 are using int8 compute as opposed to FakeQuantizeRepresentation. If we can figure out autograd story for IntRepresentation that would be great. One of the challenges, and I have heard this from other vendors, is that int tensors dont store graidents.

If some of this is not super clear, I will draft something with more details. cc: @ezyang

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [inline functorch/experimental/output_dtype.py:51]: What does this registration for CompositeExplicitAutograd do?

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [inline functorch/experimental/output_dtype.py:48]: what does this do?

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [inline functorch/experimental/output_dtype.py:30]: Why are some of these fallthrough while others have impls defined?

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [inline functorch/experimental/output_dtype.py:92]: when will we hit this? Does this exist because I can manually call this op from python code and then this line enables tracing it?

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [inline functorch/experimental/output_dtype.py:122]: isnt there equivalent with..(guard) constrcut?

File: torch/_dynamo/variables/torch.py
Comment (kimishpatel) [inline torch/_dynamo/variables/torch.py:1355]: Why does this need to be handled differently? Is it because `output_dtype_dense` does not return FakeTensor? I would have imagined the standard machinery in dynamo should take care of creating fake tensor from the output or we should just have output_dtype_dense_fake that returns fake tensor?

**Review by kimishpatel — COMMENTED** [review]
Left some comments

**Comment by kimishpatel** [comment]
Another candidate could be `cast_inputs_to_dtype` which is basically what the op is doing?

File: functorch/experimental/output_dtype.py
Comment (angelayi) [reply functorch/experimental/output_dtype.py:30]: Bc we don't have any specific things we have to implement for those keys.

File: functorch/experimental/output_dtype.py
Comment (angelayi) [reply functorch/experimental/output_dtype.py:48]: It's like a wrapper to have make_fx keep track of the results of running this function.

File: functorch/experimental/output_dtype.py
Comment (angelayi) [reply functorch/experimental/output_dtype.py:73]: Yeah, this is me avoiding implementing the autograd support for the operator just now.

File: functorch/experimental/output_dtype.py
Comment (angelayi) [reply functorch/experimental/output_dtype.py:92]: yeah

File: functorch/experimental/output_dtype.py
Comment (angelayi) [reply functorch/experimental/output_dtype.py:122]: unsure

File: torch/_dynamo/variables/torch.py
Comment (angelayi) [reply torch/_dynamo/variables/torch.py:1355]: It's because on line 1348 I get the real arguments and run it through the output_dtype operator, so the result is a real tensor. The reason why I added this if statement is because dynamo doesn't handle HigherOrderOperators out of the box; we need to special case on them.

File: functorch/experimental/output_dtype.py
Comment (jerryzh168) [reply functorch/experimental/output_dtype.py:65]: @kimishpatel thanks for writing down the details, did you get any feedback from hardware vendors how to do backward on integer compute? since it's not enough just to swap in the integer representation, we also need the computation to be differentiable in order to train.

Comment (kimishpatel) [reply functorch/experimental/output_dtype.py:65]: > feedback from hardware vendors how to do backward on integer compute?

They can do that today and some of them do, blockers for them is inability of int tensors to store gradient. But if we can enable this natively for decompositions that expressed via aten ops, that would add a lot of value.

> maybe one thing is to just use the fp version of the int8_linear operator, i.e. we convert all inputs to fp32

If we do this it will defeat the purpose of capturing compute with integer arithmetic. This is why it is somewhat non-trivial. Let me start a chat thread on this.

Comment (jerryzh168) [reply functorch/experimental/output_dtype.py:65]: > blockers for them is inability of int tensors to store gradient

is this in pytorch? I remember tried this once and it just says int Tensor doesn't have gradient, is it what you are referring to?

File: functorch/experimental/output_dtype.py
Comment (zou3519) [reply functorch/experimental/output_dtype.py:122]: Yes, the following
```py
with torch._C._ExcludeDispatchKeyGuard(...):
    ...
```
In general we should prefer these with-style python context managers to "Python RAII guards" because "Python RAII guards" are unsafe -- Python doesn't actually guarantee when your object gets deleted

**Comment by angelayi** [comment]
After offline discussion, we decided on `higher_precision` as the name, but changes some of the semantics of this op. Will pause until we figure the details out -- it'll look something like: the operator will be computed with the inputs at a general higher precision (with the inputs promoted based on some promotion table), and then the results will be cast to a user specified `output_dtype` kwarg. (look at Jerry's upcoming doc for more specific details)

File: functorch/experimental/output_dtype.py
Comment (kimishpatel) [reply functorch/experimental/output_dtype.py:122]: Aah I didnt know this about python raii guards.

File: test/functorch/test_output_dtype_op.py
Comment (kimishpatel) [reply test/functorch/test_output_dtype_op.py:18]: > Yes. How were you thinking of using the output_dtype op?

During convert step of quantization or after that when further decomposition of "quantized linear" is needed.

File: torch/_higher_order_ops/out_dtype.py
Comment (jerryzh168) [inline torch/_higher_order_ops/out_dtype.py:59]: `out_dtype` should also participate in type promotion I think

File: test/test_out_dtype_op.py
Comment (jerryzh168) [inline test/test_out_dtype_op.py:11]: can we have a test for numerics as well, i.e.

comparing
```
out_dtype(
                    torch.ops.aten.mm.default, torch.int32, x, self.weight
                )
```
and
```
# x, weight are int8 tensors
x = x.to(torch.int32)
weight = weight.to(torch.int32)
return torch.ops.aten.mm.default(x, weight)
```

**Review by jerryzh168 — COMMENTED** [review]
thanks Angela! could you also remove the third_party changes as well :)

**Comment by ezyang** [comment]
I'll let others do the main review, cc me if you need me to weigh in on something specifically.

File: test/test_out_dtype_op.py
Comment (jerryzh168) [inline test/test_out_dtype_op.py:44]: could you split the numerical test in a separate test? same for mul.Scalar, I feel it's better to separate the functional (make_fx, autograd etc.) tests and numerical tests

File: torch/_higher_order_ops/out_dtype.py
Comment (zou3519) [inline torch/_higher_order_ops/out_dtype.py:59]: Why do we promote the dtypes? I thought the semantics were that we cast the inputs to the output_dtype and run the operation.

File: torch/_higher_order_ops/out_dtype.py
Comment (zou3519) [inline torch/_higher_order_ops/out_dtype.py:84]: Should be RuntimeError or NotImplementedError. Also, we should also check the grad_mode (if we are under torch.no_grad, then we don't need to compute autograd, so we do not need to error loudly here).

File: torch/_higher_order_ops/out_dtype.py
Comment (zou3519) [inline torch/_higher_order_ops/out_dtype.py:138]: You're going to want to copy the logic from here: https://github.com/pytorch/pytorch/blob/231364fd064e08220fa78b447a817db8cddb1b28/aten/src/ATen/FunctionalizeFallbackKernel.cpp#L50

1. For each Tensor input, we need to: call sync_ on it, then use from_functional_tensor to unwrap the Tensor
2. Then, we need to do the ExcludeDispatchKeyGuard
3. Finally, we should rewrap everything in FunctionalTensorWrapper

Your code does 2 and 3, I think it is missing the sync_

File: torch/_dynamo/variables/higher_order_ops.py
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:846]: As written it looks like we're constructing real Tensors and then running the Op on them. 

Ideally we would run FakeTensor propagation on the operator to get a FakeTensor output, and then call `wrap_fx_proxy` on that.

File: torch/_dynamo/variables/higher_order_ops.py
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:848]: There should be no kwargs, right?

File: torch/_higher_order_ops/out_dtype.py
Comment (angelayi) [reply torch/_higher_order_ops/out_dtype.py:138]: I call _unwrap_all_tensors_from_functional beforehand, which I think calls sync_ [here](https://www.internalfb.com/code/fbsource/[eee9a617ea03]/fbcode/caffe2/torch/_functorch/eager_transforms.py?lines=13160). Is that enough?

Comment (zou3519) [reply torch/_higher_order_ops/out_dtype.py:138]: Yes, you're right -- sorry I missed that. This looks good to me then.

File: torch/_higher_order_ops/out_dtype.py
Comment (jerryzh168) [reply torch/_higher_order_ops/out_dtype.py:59]: the semantics actually does not really mandate how to implement it, but the current impl is what works for the use cases we are seeing right now, one example is mul.Scalar(int Tensor, float) -> int Tensor, with this we'll do computation in float in implementation and convert the result to int afterwards, which is what we expect as well

for details on semantics, please see: https://docs.google.com/document/d/12XgCdkYbVMCCWNAfIQ7U5EXKrgGmka7TqVyNn49iQpg/edit#heading=h.t725mingoxeo

File: torch/_higher_order_ops/out_dtype.py
Comment (kimishpatel) [inline torch/_higher_order_ops/out_dtype.py:93]: does this apply to ops which more than one output? If not we should probably assert that out_dtype applies to only ops which single output?

Comment (kimishpatel) [reply torch/_higher_order_ops/out_dtype.py:93]: and add tests for it.

File: test/test_out_dtype_op.py
Comment (kimishpatel) [inline test/test_out_dtype_op.py:60]: should we also add a test with aten_graph=True?

**Review by kimishpatel — COMMENTED** [review]
left a few comments

File: torch/_higher_order_ops/out_dtype.py
Comment (kimishpatel) [reply torch/_higher_order_ops/out_dtype.py]: what does sync do?

File: test/test_out_dtype_op.py
Comment (angelayi) [reply test/test_out_dtype_op.py:60]: we're using `torch._export.export` so it always returns an aten graph.

File: torch/_higher_order_ops/out_dtype.py
Comment (angelayi) [reply torch/_higher_order_ops/out_dtype.py:153]: Implementation for functorch's functionalization dispatch key.

File: torch/_higher_order_ops/out_dtype.py
Comment (zou3519) [inline torch/_higher_order_ops/out_dtype.py:53]: We shouldn't execute `op(*args)` to figure this out. We should check OpOverload's schema

File: torch/_higher_order_ops/out_dtype.py
Comment (zou3519) [inline torch/_higher_order_ops/out_dtype.py:89]: is out_dtype meant to be used with all PyTorch operators? Some PyTorch operators?

The type promotion logic may get into a problematic situation if you have an operator that intentionally takes in an index tensor, like torch.gather. If we pass it a float and a torch.long index tensor, then this logic would attempt to cast the index tensor to torch.float, which would cause the operation to error out.

The implication may be that you actually want a list of acceptable torch._ops.OpOverload that gets checked for.

File: torch/_dynamo/variables/higher_order_ops.py
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:868]: nit: real_sub_args -> fake_sub_args

File: test/test_out_dtype_op.py
Comment (zou3519) [inline test/test_out_dtype_op.py:100]: `torch.compile(f, backend="eager", fullgraph=True)`

**Review by zou3519 — COMMENTED** [review]
Code and tests look good to me. My only question is around the type promotion semantics: it seems like this would not work for all operators.

File: torch/_higher_order_ops/out_dtype.py
Comment (angelayi) [reply torch/_higher_order_ops/out_dtype.py:89]: @jerryzh168 Do we want out_dtype to be restricted to a certain set of ops now? Or are you going to followup with a PR to update with the if statement for _int8_int8_int32_mm/bmm/linear mentioned in the chat?

Comment (jerryzh168) [reply torch/_higher_order_ops/out_dtype.py:95]: @angelayi to unblock please just go with restricting it to the above ops for now

Comment (kimishpatel) [reply torch/_higher_order_ops/out_dtype.py:95]: Yeah I think we should enable this only ops that we are explicitly using it for.

File: torch/_dynamo/variables/higher_order_ops.py
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:862]: Is it possible to:
- check if there are any kwargs
- if so, then graph break (by calling `unimplemented`)?

**Review by zou3519 — APPROVED** [review]
LGTM, one minor nit

**Comment by atalman** [comment]
@pytrochbot -c nosignal -m "broke trunk win-vs2019-cpu-py3"

**Comment by atalman** [comment]
@pytrochbot revert -c nosignal -m "broke trunk win-vs2019-cpu-py3"

**Comment by atalman** [comment]
@angelayi when reopening please assign ciflow/trunk label so that we run trunk tests on it

**Comment by atalman** [comment]
@angelayi looks like windows periodic tests where broken as well, hence assigning ciflow/periodic label

**Comment by malfet** [comment]
> @pytorchbot revert -c nosignal -m "broke trunk win-vs2019-cpu-py3"

@atalman why nosignal when it clearly broke trunk? (I.e. win-vs2019-cpu-py3 ?)

**Comment by atalman** [comment]
@malfet ciflow/trunk was not assigned as a label to this PR. And when PR was merged initially there where no failures, if you look at the failures now, these are recent failures after reopening and assigning ciflow/trunk label

**Pull request closed without merging**
